src/cholesky/lkjcholesky.jl

```diff
@@ -0,0 +1,241 @@
+        return (c - i) * log(factors[di])
+    end
+    return logp
+end
+
+function logpdf(d::LKJCholesky, R::Cholesky)
+    insupport(d, R) || throw(ArgumentError("Provided point is not in the support."))
+    return _logpdf(d, R)
+end
+
+_logpdf(d::LKJCholesky, R::Cholesky) = logkernel(d, R) + d.logc0
+
+pdf(d::LKJCholesky, R::Cholesky) = exp(logpdf(d, R))
+
+loglikelihood(d::LKJCholesky, R::Cholesky) = logpdf(d, R)
+function loglikelihood(d::LKJCholesky, Rs::AbstractArray{<:Cholesky})
+    return sum(R -> logpdf(d, R), Rs)
+end
+
+#  -----------------------------------------------------------------------------
+#  Sampling
+#  -----------------------------------------------------------------------------
+
+function rand(rng::AbstractRNG, d::LKJCholesky)
+    factors = Matrix{eltype(d)}(undef, size(d))
+    R = Cholesky(factors, d.uplo, 0)
+    return _lkj_cholesky_onion_sampler!(rng, d, R)
+end
+function rand(rng::AbstractRNG, d::LKJCholesky, dims::Dims)
+    p = dim(d)
+    uplo = d.uplo
+    T = eltype(d)
+    TM = Matrix{T}
+    Rs = Array{Cholesky{T,TM}}(undef, dims)
+    for i in eachindex(Rs)
+        factors = TM(undef, p, p)
+        Rs[i] = R = Cholesky(factors, uplo, 0)
+        _lkj_cholesky_onion_sampler!(rng, d, R)
+    end
+    return Rs
+end
+
+rand!(d::LKJCholesky, R::Cholesky) = rand!(GLOBAL_RNG, d, R)
+rand!(rng::AbstractRNG, d::LKJCholesky, R::Cholesky) = _lkj_cholesky_onion_sampler!(rng, d, R)
+
+function rand!(rng::AbstractRNG, d::LKJCholesky, Rs::AbstractArray{<:Cholesky{T,TM}}, allocate::Bool) where {T,TM}
+    p = dim(d)
+    uplo = d.uplo
+    if allocate
+        for i in eachindex(Rs)
+            Rs[i] = _lkj_cholesky_onion_sampler!(rng, d, Cholesky(TM(undef, p, p), uplo, 0))
+        end
+    else
+        for i in eachindex(Rs)
+            _lkj_cholesky_onion_sampler!(rng, d, Rs[i])
+        end
+    end
+    return Rs
+end
+function rand!(rng::AbstractRNG, d::LKJCholesky, Rs::AbstractArray{<:Cholesky{<:Real}})
+    allocate = any(!isassigned(Rs, i) for i in eachindex(Rs)) || any(R -> size(R, 1) != dim(d), Rs)
+    return rand!(rng, d, Rs, allocate)
+end
+
+#
+# onion method
+#
+
+function _lkj_cholesky_onion_sampler!(rng::AbstractRNG, d::LKJCholesky, R::Cholesky)
+    TTri = R.uplo === 'U' ? UpperTriangular : LowerTriangular
+    _lkj_cholesky_onion_tri!(rng, R.factors, d.d, d.η, TTri)
+    return R
+end
+
+function _lkj_cholesky_onion_tri!(
+    rng::AbstractRNG,
+    A::AbstractMatrix,
+    d::Integer,
+    η::Real,
+    ::Type{TTri},
+) where {TTri<:LinearAlgebra.AbstractTriangular}
+    # Section 3.2 in LKJ (2009 JMA)
+    # reformulated to incrementally construct Cholesky factor as mentioned in Section 5
+    # equivalent steps in algorithm in reference are marked.
+    @assert size(A) == (d, d)
+    A[1, 1] = 1
+    d > 1 || return R
+    β = η + (d - 2)//2
+    #  1. Initialization
+    w0 = 2 * rand(rng, Beta(β, β)) - 1
+    @inbounds if TTri <: LowerTriangular
```

Comment:
Maybe just fall back to `_lkj_cholesky_onion_tri!(rng, A', d, eta, TTri)` in one of both cases?

Reply:
I introduced the type parameter because operating on the transpose increased the runtime by ~30%. This way we do the fast thing for both triangles with almost no code duplication.